packages/cli/src/lib/setup/setupSetup.ts

```diff
@@ -9,36 +9,72 @@
 
 'use strict';
 
-const fs = require('fs-extra');
-const path = require('path');
-const { tools, EXIT_CODES } = require('@iobroker/js-controller-common');
-const dbTools = require('@iobroker/js-controller-common-db').tools;
-const Backup = require('@iobroker/js-controller-cli').setupBackup;
-const deepClone = require('deep-clone');
-const pluginInfos = require('./pluginInfos');
+import type {
+    CleanDatabase,
+    DbConnectAsync,
+    IobrokerJSON,
+    IoPackage,
+    ProcessExit,
+    ResetDbConnect,
+    RestartController
+} from '../_Types';
+import type { Client as StatesRedisClient } from '@iobroker/db-states-redis';
+import type { Client as ObjectsRedisClient } from '@iobroker/db-objects-redis';
+
+import fs from 'fs-extra';
+import path from 'path';
+import { EXIT_CODES, tools } from '@iobroker/js-controller-common';
+import { tools as dbTools } from '@iobroker/js-controller-common-db';
+import { BackupRestore } from './setupBackup';
+import crypto from 'crypto';
+import deepClone from 'deep-clone';
+import * as pluginInfos from './pluginInfos';
+import rl from 'readline-sync';
+import os from 'os';
 
 const COLOR_RED = '\x1b[31m';
 const COLOR_YELLOW = '\x1b[33m';
 const COLOR_RESET = '\x1b[0m';
 const COLOR_GREEN = '\x1b[32m';
 const CONTROLLER_DIR = tools.getControllerDir();
 
-/** @class */
-function Setup(options) {
-    options = options || {};
+export interface CLISetupOptions {
+    dbConnectAsync: DbConnectAsync;
+    cleanDatabase: CleanDatabase;
+    processExit: ProcessExit;
+    params: Record<string, any>;
+    restartController: RestartController;
+    resetDbConnect: ResetDbConnect;
+}
 
-    const processExit = options.processExit;
-    const dbConnect = options.dbConnect;
-    const params = options.params;
-    const cleanDatabase = options.cleanDatabase;
-    const resetDbConnect = options.resetDbConnect;
-    const restartController = options.restartController;
-    let objects;
-    let states;
+type ConfigObject = ioBroker.OtherObject & { type: 'config' };
+
+export class Setup {
+    private readonly options: CLISetupOptions;
+    private readonly processExit: ProcessExit;
+    private states: StatesRedisClient | undefined;
+    private objects: ObjectsRedisClient | undefined;
+    private readonly dbConnectAsync: DbConnectAsync;
+    private readonly params: Record<string, any>;
+    private readonly cleanDatabase: CleanDatabase;
+    private readonly restartController: RestartController;
+    private readonly resetDbConnect: ResetDbConnect;
+
+    constructor(options: CLISetupOptions) {
+        this.options = options || {};
```

Comment:
```suggestion
    constructor(options: CLISetupOptions = {}) {
```

Reply:
Good call. I guess it is more an error, that options is not optional, the fallback was legacy code which I just ported and if the fallback would ever apply, the controller would crash, as e.g. `this.processExit` calls are not guarded. I will remove the obsolete fallback.